File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        if(ceil(sqrt(n))==floor(sqrt(n))): return 1
        while(n%4 == 0): n//=4
        if(n%8==7): return 4
        i =1
        while(i*i <=n):
            b = floor(sqrt(n-i*i))
            if(b*b == n- i*i): return 2
            i=i+1 
        return 3

File: 0279-perfect-squares/0279-perfect-squares.cpp
//Legendre's 3-square theorem
class Solution {
public:
    int numSquares(int n) {
        if(ceil(sqrt(n))==floor(sqrt(n)))
            return 1;
        while(n%4==0)   
            n/=4;
        if(n%8==7)      
            return 4;
        for(int i=1;i*i<=n;++i)
        {
            int b= sqrt(n-i*i);
            if(b*b==(n-i*i))
                return 2;
        }
        return 3;
    }
};